Defer resource cleanup in integ test

SoManyHs opened the pull request:
We were seeing services and empty clusters being left over when UpdateService was being called immediately after CreateService, due to eventual read consistency issues in ECS, mostly on the `TestCreateClusterWithFargateService` integration test.. This would cause the UpdateService call to fail, which meant the cleanup actions (DeleteService, DeleteCluster) were not getting called. This ensures they will be even if intermediate calls inside the test function fail.

----
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

efekarakus commented on ecs-cli/integ/e2e/fargate_service_test.go:
```diff
@@ -41,19 +41,17 @@ func TestCreateClusterWithFargateService(t *testing.T) {
 
+	// Ensure cleanup of service and cluster
+	defer cmd.TestDown(t, conf)
+	defer cmd.TestServiceDown(t, project)
```
the `defer` is called before `TestServiceUp` so I don't know how that will behave in case `ServiceUp` fails :P 

Should we move `defer cmd.TestDown(t, conf)` to l.36 and `defer cmd.TestServiceDown(t, project)` to l.50 instead?

SoManyHs replied:
good call. fixed!

SoManyHs commented:
Manual testing: 

Force a failure in  `[TestCreateClusterWithFargateService](https://github.com/aws/amazon-ecs-cli/blob/64588694ea7d40e45019f28da37fa27696e55d8f/ecs-cli/integ/e2e/fargate_service_test.go#L50-L51)` by changing the assertion on number of 
running tasks (exported `AWS_REGION=us-west-2` and `CODEBUILD_BUILD_ID=12345` env vars so the test would work):
```
	// Increase the number of running tasks
	cmd.TestServiceScale(t, project, 2)
	cmd.TestServicePs(t, project, 3)
```

Running that test before any other changes:
```
--- FAIL: TestCreateClusterWithFargateService (307.74s)
    configure.go:42: Created config 12345-fargate-tutorial-config
    up.go:67: Created cluster 12345-fargate-tutorial-cluster in stack amazon-ecs-cli-setup-12345-fargate-tutorial-cluster
    fargate_service_test.go:84: Created /var/folders/13/3nq2m6fj4x7674z_zjn177yw4cfpw_/T/docker-compose-966772162.yml successfully
    fargate_service_test.go:115: Created /var/folders/13/3nq2m6fj4x7674z_zjn177yw4cfpw_/T/ecs-params-575228473.yml successfully
    compose.go:102: Created service with name e2e-fargate-test-service
    compose.go:280: Container is RUNNING: 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN
    compose.go:113: Project e2e-fargate-test-service has 1 running containers
    compose.go:173: Scaled the service e2e-fargate-test-service to 2 tasks
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:31:10.011422 -0800 PST m=+178.621713999, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:31:26.335124 -0800 PST m=+194.944926097, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:31:42.686887 -0800 PST m=+211.296198431, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:31:59.000323 -0800 PST m=+227.609145481, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:32:15.257954 -0800 PST m=+243.852792413, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:32:31.648343 -0800 PST m=+260.236600788, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:32:47.89566 -0800 PST m=+276.481325091, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/4fae9657f1ee47b0b4d070d62cc49946/wordpress  RUNNING  3.16.21.171:80->80/tcp     e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/de785b3d68f3473bb9d88ddd09618720/wordpress  RUNNING  18.223.239.125:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:33:04.157112 -0800 PST m=+292.741551108, sleeping for 15s
    require.go:1159: 
        	Error Trace:	compose.go:112
        	Error:      	Should be true
        	Test:       	TestCreateClusterWithFargateService
        	Messages:   	Failed to get RUNNING containers
FAIL
FAIL	github.com/aws/amazon-ecs-cli/ecs-cli/integ/e2e	308.948s
```

**NOTE: not quite the same failure as what we've been seeing in codebuild, but the goal is to simulate the resource cleanup triggering for any failure.**

Sure enough, cluster and stack are still there:
```
~/go/src/github.com/aws/amazon-ecs-cli(master)$ aws ecs list-clusters
{
    "clusterArns": [
        "arn:aws:ecs:us-east-2:xxxxxxxxxxx:cluster/12345-fargate-tutorial-cluster"
    ]
}

{
    "StackSummaries": [
        {
            "StackId": "arn:aws:cloudformation:us-east-2:xxxxxxxx:stack/amazon-ecs-cli-setup-12345-fargate-tutorial-cluster/f4bd1380-546a-11ea-9744-0294f7a52ca2", 
            "DriftInformation": {
                "StackDriftStatus": "NOT_CHECKED"
            }, 
            "TemplateDescription": "AWS CloudFormation template to create resources required to run tasks on an ECS cluster.", 
            "CreationTime": "2020-02-21T05:28:13.959Z", 
            "StackName": "amazon-ecs-cli-setup-12345-fargate-tutorial-cluster", 
            "StackStatus": "CREATE_COMPLETE"
        }, 
...
```

With the deferred cluster and service deletion:

```
=== RUN   TestCreateClusterWithFargateService
--- FAIL: TestCreateClusterWithFargateService (387.40s)
    configure.go:42: Created config 12345-fargate-tutorial-config
    up.go:67: Created cluster 12345-fargate-tutorial-cluster in stack amazon-ecs-cli-setup-12345-fargate-tutorial-cluster
    fargate_service_test.go:86: Created /var/folders/13/3nq2m6fj4x7674z_zjn177yw4cfpw_/T/docker-compose-945024723.yml successfully
    fargate_service_test.go:117: Created /var/folders/13/3nq2m6fj4x7674z_zjn177yw4cfpw_/T/ecs-params-072698902.yml successfully
    compose.go:102: Created service with name e2e-fargate-test-service
    compose.go:280: Container is RUNNING: 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN
    compose.go:113: Project e2e-fargate-test-service has 1 running containers
    compose.go:173: Scaled the service e2e-fargate-test-service to 2 tasks
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:56:37.404285 -0800 PST m=+164.072286758, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:56:53.613256 -0800 PST m=+180.280771541, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:57:09.816659 -0800 PST m=+196.483688185, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:57:26.092326 -0800 PST m=+212.758867189, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:57:42.306339 -0800 PST m=+228.972393371, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:57:58.503095 -0800 PST m=+245.168663378, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:58:14.743798 -0800 PST m=+261.408879670, sleeping for 15s
    compose.go:270: Wanted = 3, got = 2 running containers
    compose.go:271: Current running containers = [12345-fargate-tutorial-cluster/3970f5a1d1ef4902b83424ac0afae9e9/wordpress  RUNNING  13.58.58.226:80->80/tcp   e2e-fargate-test-service:1  UNKNOWN 12345-fargate-tutorial-cluster/9c503666ac8a4d91a48ff3ac98053bb1/wordpress  RUNNING  52.14.226.236:80->80/tcp  e2e-fargate-test-service:1  UNKNOWN]
    runner.go:98: Current timestamp=2020-02-20 21:58:30.934039 -0800 PST m=+277.598635319, sleeping for 15s
    require.go:1159: 
        	Error Trace:	compose.go:112
        	Error:      	Should be true
        	Test:       	TestCreateClusterWithFargateService
        	Messages:   	Failed to get RUNNING containers
    compose.go:232: Deleted service e2e-fargate-test-service
    down.go:49: Deleted cluster 12345-fargate-tutorial-cluster
    cfn.go:50: Success: stack amazon-ecs-cli-setup-12345-fargate-tutorial-cluster does not exist
    down.go:52: Deleted stack amazon-ecs-cli-setup-12345-fargate-tutorial-cluster
FAIL
FAIL	github.com/aws/amazon-ecs-cli/ecs-cli/integ/e2e	387.504s
FAIL
```

Stack and cluster are deleted.
```
~/go/src/github.com/aws/amazon-ecs-cli(fix-integ)$ aws ecs list-clusters
{
    "clusterArns": [
    ]
}
```

bvtujo commented:
🚫🧟‍♂️🚀
